libcfnet/tls_client.c

```diff
@@ -206,14 +206,30 @@ int TLSClientIdentificationDialog(ConnectionInfo *conn_info,
     ProtocolVersion wanted_version;
     if (conn_info->protocol == CF_PROTOCOL_UNDEFINED)
     {
-        /* TODO parse CFE_v%d received and use that version if it's lower. */
         wanted_version = CF_PROTOCOL_LATEST;
     }
     else
     {
         wanted_version = conn_info->protocol;
     }
 
+    const ProtocolVersion received_version = ParseProtocolVersionNetwork(line);
```

Comment:
What populates the `conn_info->protocol` which is compared above?

Reply:
This client code is used in many different scenarios. For filecopy it is based on policy:

```
    ConnectionFlags flags = {
        .protocol_version = DecideProtocol(ctx, fc->protocol_version),
        .cache_connection = !background,
        .force_ipv4 = fc->force_ipv4,
        .trust_server = fc->trustkey,
        .off_the_record = false
    };
    [...]
    AgentConnection *conn = ServerConnection(servername, port, conntimeout, flags, &err);
```

In other places, like cf-hub it will always attempt latest:

```
    ConnectionFlags flags = { .protocol_version = CF_PROTOCOL_LATEST };
    return ServerConnection(ip_address, port, 30, flags, err_code);
```